**Linux** [Build 8835](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=80389&buildTypeId=WFPR) outcome was **FAILURE** using a merge of bea9265d9d9c53639d2ec62bd7d6b2ad2525ae72
Summary: Tests failed: 1 (1 new), passed: 3267, ignored: 363 Build time: 00:51:19
### Failed tests
```
org.jboss.as.test.integration.messaging.mgmt.JMSQueueManagementTestCase.testMessageRemoval: java.lang.AssertionError: expected:<2> but was:<0>
	at org.jboss.as.test.integration.messaging.mgmt.JMSQueueManagementTestCase.testMessageRemoval(JMSQueueManagementTestCase.java:226)
------- Stdout: -------
 [0m12:21:26,800 INFO  [org.apache.activemq.artemis.core.server] (ServerService Thread Pool -- 22) AMQ221003: trying to deploy queue jms.queue.JMSQueueManagementTestCase1448623285687
 [0m12:21:26,805 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "add",
    "address" => [
        ("subsystem" => "messaging-activemq"),
        ("server" => "default"),
        ("jms-queue" => "JMSQueueManagementTestCase1448623285687")
    ],
    "entries" => ["java:jboss/exported/queue/JMSQueueManagementTestCase1448623285687"]
}
 [0m12:21:26,806 INFO  [org.apache.activemq.artemis.core.server] (ServerService Thread Pool -- 22) AMQ221003: trying to deploy queue jms.queue.JMSQueueManagementTestCaseother1448623285687
 [0m12:21:26,811 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "add",
    "address" => [
        ("subsystem" => "messaging-activemq"),
        ("server" => "default"),
        ("jms-queue" => "JMSQueueManagementTestCaseother1448623285687")
    ],
    "entries" => ["java:jboss/exported/queue/JMSQueueManagementTestCaseother1448623285687"]
}
12:21:26,894 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "remove",
    "address" => [
        ("subsystem" => "messaging-activemq"),
        ("server" => "default"),
        ("jms-queue" => "JMSQueueManagementTestCase1448623285687")
    ]
}
12:21:26,923 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "remove",
    "address" => [
        ("subsystem" => "messaging-activemq"),
        ("server" => "default"),
        ("jms-queue" => "JMSQueueManagementTestCaseother1448623285687")
    ]
}


```

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/CallProtectedEjbServlet.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/CallProtectedEjbServlet.java
@@ -1,0 +1,116 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat Middleware LLC, and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.test.integration.web.security.runas;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.annotation.security.DeclareRoles;
+import javax.ejb.EJB;
+import javax.naming.NamingException;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.jboss.logging.Logger;
+
+/**
+ * Protected servlet which calls protected EJB method {@link Hello#sayHello()}.
+ *
+ * @author olukas
+ */
+@WebServlet(CallProtectedEjbServlet.SERVLET_PATH)
+@DeclareRoles({HelloBean.AUTHORIZED_ROLE, HelloBean.NOT_AUTHORIZED_ROLE})
+public class CallProtectedEjbServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOGGER = Logger.getLogger(CallProtectedEjbServlet.class);
+
+    public static final String SERVLET_PATH = "/CallProtectedEjbServlet";
+    public static final String FILE_PARAM = "file";
+    public static final String INIT_METHOD_PASS = "init method passed.";
+    public static final String INIT_METHOD_NOT_PASS = "init method did not pass.";
+    public static final String DOGET_METHOD_PASS = "doGet method passed.";
+    public static final String DOGET_METHOD_NOT_PASS = "doGet method did not pass.";
+    public static final String DESTROY_METHOD_PASS = "destroy method passed.";
+    public static final String DESTROY_METHOD_NOT_PASS = "destroy method did not pass.";
+
+    private static String initMethodPass;
+    private String filePath;
+
+    @EJB
+    private Hello ejb;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        LOGGER.debug("Calling EJB protected method in doGet() method.");
+
+        resp.setContentType("text/plain");
+        filePath = req.getParameter(FILE_PARAM);
+
+        try (PrintWriter writer = resp.getWriter()) {
+            writer.println(initMethodPass);
+            try {
+                String callProtectedEJB = callProtectedEJB();
+                writer.println(DOGET_METHOD_PASS + callProtectedEJB);
+            } catch (Exception e) {
+                writer.println(DOGET_METHOD_NOT_PASS);
+                LOGGER.debug(e);
+            }
+        }
+    }
+
+    @Override
+    public void init() throws ServletException {
+        LOGGER.debug("Calling EJB protected method in init() method.");
+        try {
+            String callProtectedEJB = callProtectedEJB();
+            initMethodPass = INIT_METHOD_PASS + callProtectedEJB;
+        } catch (Exception e) {
+            initMethodPass = INIT_METHOD_NOT_PASS;
+            LOGGER.debug(e);
+        }
+    }
+
+    @Override
+    public void destroy() {
+        LOGGER.debug("Calling EJB protected method in init() method.");
+        try (PrintWriter writer = new PrintWriter(filePath)) {
+            try {
+                String callProtectedEJB = callProtectedEJB();
+                writer.println(DESTROY_METHOD_PASS + callProtectedEJB);
+            } catch (Exception e) {
+                writer.println(DESTROY_METHOD_NOT_PASS);
+                LOGGER.debug(e);
+            }
+        } catch (FileNotFoundException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    protected String callProtectedEJB() throws NamingException {
+        return ejb.sayHello();
+    }
+
+}

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/Hello.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/Hello.java
@@ -1,0 +1,32 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.test.integration.web.security.runas;
+
+/**
+ *
+ * @author olukas
+ */
+public interface Hello {
+
+    String sayHello();
+
+}

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/HelloBean.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/HelloBean.java
@@ -1,0 +1,49 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.test.integration.web.security.runas;
+
+import javax.annotation.security.DeclareRoles;
+import javax.annotation.security.RolesAllowed;
+import javax.ejb.Remote;
+import javax.ejb.Stateless;
+
+/**
+ *
+ * @author olukas
+ */
+@DeclareRoles({HelloBean.AUTHORIZED_ROLE, HelloBean.NOT_AUTHORIZED_ROLE})
+@Stateless
+@RolesAllowed(HelloBean.AUTHORIZED_ROLE)
+@Remote(Hello.class)
+public class HelloBean implements Hello {
+
+    public static final String AUTHORIZED_ROLE = "Admin";
+    public static final String NOT_AUTHORIZED_ROLE = "User";
+
+    public static final String HELLO = "Hello!";
+
+    @Override
+    public String sayHello() {
+        return HELLO;
+    }
+
+}

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/RunAsAdminServlet.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/RunAsAdminServlet.java
@@ -1,0 +1,42 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat Middleware LLC, and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.test.integration.web.security.runas;
+
+import javax.annotation.security.DeclareRoles;
+import javax.annotation.security.RunAs;
+import javax.servlet.annotation.WebServlet;
+
+/**
+ * RunAs annotated servlet which calls protected EJB method {@link Hello#sayHello()}.
+ *
+ * @author olukas
+ */
+@WebServlet(RunAsAdminServlet.SERVLET_PATH)
+@DeclareRoles({HelloBean.AUTHORIZED_ROLE, HelloBean.NOT_AUTHORIZED_ROLE})
+@RunAs(HelloBean.AUTHORIZED_ROLE)
+public class RunAsAdminServlet extends CallProtectedEjbServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    public static final String SERVLET_PATH = "/RunAsAdminServlet";
+
+}

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/RunAsUserServlet.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/RunAsUserServlet.java
@@ -1,0 +1,43 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat Middleware LLC, and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.test.integration.web.security.runas;
+
+import javax.annotation.security.DeclareRoles;
+import javax.annotation.security.RunAs;
+import javax.servlet.annotation.WebServlet;
+
+/**
+ * RunAs annotated servlet which calls protected EJB method {@link Hello#sayHello()}. Role used in RunAs in not correct role for
+ * protected EJB method.
+ *
+ * @author olukas
+ */
+@WebServlet(RunAsUserServlet.SERVLET_PATH)
+@DeclareRoles({HelloBean.AUTHORIZED_ROLE, HelloBean.NOT_AUTHORIZED_ROLE})
+@RunAs(HelloBean.NOT_AUTHORIZED_ROLE)
+public class RunAsUserServlet extends CallProtectedEjbServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    public static final String SERVLET_PATH = "/RunAsUserServlet";
+
+}

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/ServletRunAsTestCase.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/ServletRunAsTestCase.java
@@ -1,0 +1,283 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat Middleware LLC, and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.test.integration.web.security.runas;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import static java.net.HttpURLConnection.HTTP_OK;
+import java.net.URL;
+import org.apache.commons.io.FileUtils;
+import org.jboss.arquillian.container.test.api.Deployer;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.container.test.api.OperateOnDeployment;
+import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.arquillian.junit.InSequence;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.arquillian.container.ManagementClient;
+import org.jboss.as.test.integration.security.common.Utils;
+import static org.jboss.as.test.integration.web.security.runas.CallProtectedEjbServlet.DESTROY_METHOD_NOT_PASS;
+import static org.jboss.as.test.integration.web.security.runas.CallProtectedEjbServlet.DESTROY_METHOD_PASS;
+import static org.jboss.as.test.integration.web.security.runas.CallProtectedEjbServlet.DOGET_METHOD_NOT_PASS;
+import static org.jboss.as.test.integration.web.security.runas.CallProtectedEjbServlet.DOGET_METHOD_PASS;
+import static org.jboss.as.test.integration.web.security.runas.CallProtectedEjbServlet.INIT_METHOD_NOT_PASS;
+import static org.jboss.as.test.integration.web.security.runas.CallProtectedEjbServlet.INIT_METHOD_PASS;
+import org.jboss.as.test.shared.ServerReload;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import static org.junit.Assert.assertTrue;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * Test case for RunAs in Servlet. It tests secured EJB invocation in HttpServlet init(), doGet() and destroy() methods.
+ *
+ * @author olukas
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+public class ServletRunAsTestCase {
+
+    private static final String DEPLOYMENT = "deployment";
+
+    private static String correctRoleResult;
+    private static String incorrectRoleResult;
+
+    private static final File WORK_DIR = new File("ServletRunAsTestCase-" + System.currentTimeMillis());
+    private static final File CORRECT_ROLE_AND_STOP_SERVER = new File(WORK_DIR, "correctRoleAndStopServer.log");
+    private static final File INCORRECT_ROLE_AND_STOP_SERVER = new File(WORK_DIR, "incorrectRoleAndStopServer.log");
+    private static final File CORRECT_ROLE_AND_UNDEPLOY = new File(WORK_DIR, "correctRoleAndUndeploy.log");
+    private static final File INCORRECT_ROLE_AND_UNDEPLOY = new File(WORK_DIR, "incorrectRoleAndUndeploy.log");
+
+    @ArquillianResource
+    private ManagementClient managementClient;
+
+    @ArquillianResource
+    Deployer deployer;
+
+    @Test
+    @InSequence(Integer.MIN_VALUE)
+    public void initialize() throws Exception {
+        deployer.deploy(DEPLOYMENT);
+        WORK_DIR.mkdirs();
+    }
+
+    /**
+     * Access Servlet which uses RunAs with correct role needed for secured EJB invocation.
+     *
+     * This method will run init() and doGet() method and stores results.
+     *
+     * @param url
+     * @throws Exception
+     */
+    @Test
+    @InSequence(1)
+    @OperateOnDeployment(DEPLOYMENT)
+    public void runServletWithCorrectRole(@ArquillianResource URL url) throws Exception {
+        final URL servletUrl = new URL(url.toExternalForm() + RunAsAdminServlet.SERVLET_PATH.substring(1) + "?"
+                + Utils.encodeQueryParam(CallProtectedEjbServlet.FILE_PARAM, CORRECT_ROLE_AND_UNDEPLOY.getAbsolutePath()));
+        correctRoleResult = Utils.makeCall(servletUrl.toURI(), HTTP_OK);
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with correct role needed for secured EJB invocation can correctly invoked that EJB
+     * method in HttpServlet.init() method.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(2)
+    public void checkInitMethodWithCorrectRole() throws Exception {
+        assertTrue("EJB invocation failed in init() method of Servlet which uses RunAs with correct role.",
+                correctRoleResult.contains(INIT_METHOD_PASS + HelloBean.HELLO));
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with correct role needed for secured EJB invocation can correctly invoked that EJB
+     * method in HttpServlet.doGet() method.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(3)
+    public void checkDoGetMethodWithCorrectRole() throws Exception {
+        assertTrue("EJB invocation failed in doGet() method of Servlet which uses RunAs with correct role.",
+                correctRoleResult.contains(DOGET_METHOD_PASS + HelloBean.HELLO));
+    }
+
+    /**
+     * Access Servlet which uses RunAs with different role than in needed for secured EJB invocation.
+     *
+     * This method will run init() and doGet() method and stores results.
+     *
+     * @param url
+     * @throws Exception
+     */
+    @Test
+    @InSequence(4)
+    @OperateOnDeployment(DEPLOYMENT)
+    public void runServletWithIncorrectRole(@ArquillianResource URL url) throws Exception {
+        final URL servletUrl = new URL(url.toExternalForm() + RunAsUserServlet.SERVLET_PATH.substring(1) + "?"
+                + Utils.encodeQueryParam(CallProtectedEjbServlet.FILE_PARAM, INCORRECT_ROLE_AND_UNDEPLOY.getAbsolutePath()));
+        incorrectRoleResult = Utils.makeCall(servletUrl.toURI(), HTTP_OK);
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with different role than in needed for secured EJB invocation cannot correctly
+     * invoked that EJB method in HttpServlet.init() method.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(5)
+    public void checkInitMethodWithIncorrectRole() throws Exception {
+        assertTrue("EJB invocation did not failed in init() method of Servlet which uses RunAs with incorrect role.",
+                incorrectRoleResult.contains(INIT_METHOD_NOT_PASS));
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with different role than in needed for secured EJB invocation cannot correctly
+     * invoked that EJB method in HttpServlet.doGet() method.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(6)
+    public void checkDoGetMethodWithIncorrectRole() throws Exception {
+        assertTrue("EJB invocation did not failed in doGet() method of Servlet which uses RunAs with incorrect role.",
+                incorrectRoleResult.contains(DOGET_METHOD_NOT_PASS));
+    }
+
+    /**
+     * Stop server for invoking HttpServlet.destroy() method during undeploying.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(7)
+    public void runDestroyMethodInUndeploying() throws Exception {
+        deployer.undeploy(DEPLOYMENT);
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with correct role needed for secured EJB invocation can correctly invoked that EJB
+     * method in HttpServlet.destroy() method during undeploying application.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(8)
+    public void checkDestroyInUndeployingMethodWithCorrectRole() throws Exception {
+        assertTrue("EJB invocation failed in destroy() method of Servlet which uses RunAs with correct role during undeploying.",
+                readFirstLineOfFile(CORRECT_ROLE_AND_UNDEPLOY).contains(DESTROY_METHOD_PASS + HelloBean.HELLO));
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with different role than in needed for secured EJB invocation cannot correctly
+     * invoked that EJB method in HttpServlet.destroy() method during undeploying application.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(9)
+    public void checkDestroyInUndeployingMethodWithIncorrectRole() throws Exception {
+        assertTrue("EJB invocation did not failed in destroy() method of Servlet which uses RunAs with incorrect role during "
+                + "undeploying.",
+                readFirstLineOfFile(INCORRECT_ROLE_AND_UNDEPLOY).contains(DESTROY_METHOD_NOT_PASS));
+    }
+
+    /**
+     * Restart server for invoking HttpServlet.destroy() method during stopping server. It also hit Servlet for initialization
+     * of Servlet before server is restarted.
+     *
+     * @param url
+     * @throws Exception
+     */
+    @Test
+    @InSequence(10)
+    @OperateOnDeployment(DEPLOYMENT)
+    public void runDestroyMethodInStopServer(@ArquillianResource URL url) throws Exception {
+        deployer.deploy(DEPLOYMENT);
+
+        URL servletUrl = new URL(url.toExternalForm() + RunAsAdminServlet.SERVLET_PATH.substring(1) + "?"
+                + Utils.encodeQueryParam(CallProtectedEjbServlet.FILE_PARAM, CORRECT_ROLE_AND_STOP_SERVER.getAbsolutePath()));
+        Utils.makeCall(servletUrl.toURI(), HTTP_OK);
+        servletUrl = new URL(url.toExternalForm() + RunAsUserServlet.SERVLET_PATH.substring(1) + "?"
+                + Utils.encodeQueryParam(CallProtectedEjbServlet.FILE_PARAM, INCORRECT_ROLE_AND_STOP_SERVER.getAbsolutePath()));
+        Utils.makeCall(servletUrl.toURI(), HTTP_OK);
+
+        ServerReload.executeReloadAndWaitForCompletion(managementClient.getControllerClient(), 50000);
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with correct role needed for secured EJB invocation can correctly invoked that EJB
+     * method in HttpServlet.destroy() method during stopping server.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(11)
+    public void checkDestroyMethodInStopServerWithCorrectRole() throws Exception {
+        assertTrue("EJB invocation failed in destroy() method of Servlet which uses RunAs with correct role during stopping "
+                + "server.",
+                readFirstLineOfFile(CORRECT_ROLE_AND_STOP_SERVER).contains(DESTROY_METHOD_PASS + HelloBean.HELLO));
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with different role than in needed for secured EJB invocation cannot correctly
+     * invoked that EJB method in HttpServlet.destroy() method during stopping server.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(12)
+    public void checkDestroyMethodInStopServerWithIncorrectRole() throws Exception {
+        assertTrue("EJB invocation did not failed in destroy() method of Servlet which uses RunAs with incorrect role "
+                + "during stopping server.",
+                readFirstLineOfFile(INCORRECT_ROLE_AND_STOP_SERVER).contains(DESTROY_METHOD_NOT_PASS));
+    }
+
+    @Test
+    @InSequence(Integer.MAX_VALUE)
+    public void finish() throws Exception {
+        deployer.undeploy(DEPLOYMENT);
+        FileUtils.deleteQuietly(WORK_DIR);
+    }
+
+    @Deployment(name = DEPLOYMENT, managed = false, testable = false)
+    public static WebArchive deployment() throws Exception {
+        WebArchive war = ShrinkWrap.create(WebArchive.class, "servlet-runas.war");
+        war.addClasses(CallProtectedEjbServlet.class, RunAsAdminServlet.class, RunAsUserServlet.class, Hello.class,
+                HelloBean.class);
+        return war;
+    }
+
+    private String readFirstLineOfFile(File file) throws FileNotFoundException, IOException {
+        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
+            return br.readLine();
+        }
+    }
+
+}

